File: shopGunSdk/src/main/java/com/tjek/sdk/publicationviewer/paged/zoomlayout/ZoomOnDoubleTapListener.java
package com.tjek.sdk.publicationviewer.paged.zoomlayout;

public class ZoomOnDoubleTapListener implements ZoomLayoutInterface.OnDoubleTapListener {

    private boolean mThreeStep = false;

    public ZoomOnDoubleTapListener(boolean threeStep) {
        mThreeStep = threeStep;
    }

    @Override
    public boolean onDoubleTap(ZoomLayout view, TapInfo info) {
        try {
            if (mThreeStep) {
                threeStep(view, info.getAbsoluteX(), info.getAbsoluteY());
            } else {
                twoStep(view, info.getAbsoluteX(), info.getAbsoluteY());
            }
        } catch (ArrayIndexOutOfBoundsException e) {
            // Can sometimes happen when getX() and getY() is called
        }
        return true;
    }

    private void twoStep(ZoomLayout view, float x, float y) {
        if (view.getScale() > view.getMinScale()) {
            view.setScale(view.getMinScale(), true);
        } else {
            view.setScale(view.getMaxScale(), x, y, true);
        }
    }

    private void threeStep(ZoomLayout view, float x, float y) {
        float scale = view.getScale();
        float medium = view.getMinScale() + ((view.getMaxScale() - view.getMinScale()) * 0.3f);
        if (scale < medium) {
            view.setScale(medium, x, y, true);
        } else if (scale >= medium && scale < view.getMaxScale()) {
            view.setScale(view.getMaxScale(), x, y, true);
        } else {
            view.setScale(view.getMinScale(), true);
        }
    }

}

File: shopGunSdk/src/main/java/com/tjek/sdk/publicationviewer/paged/zoomlayout/ZoomOnDoubleTapListener.kt
package com.tjek.sdk.publicationviewer.paged.zoomlayout

class ZoomOnDoubleTapListener(threeStep: Boolean) : EventListener {

    private var mThreeStep = threeStep

    override fun onEvent(event: Event): Boolean {
        return when (event) {
            is Event.DoubleTap -> onDoubleTap(event.view, event.info)
            else -> false
        }
    }

    private fun onDoubleTap(view: ZoomLayout, info: TapInfo): Boolean {
        try {
            if (mThreeStep) {
                threeStep(view, info.absoluteX, info.absoluteY)
            } else {
                twoStep(view, info.absoluteX, info.absoluteY)
            }
        } catch (e: ArrayIndexOutOfBoundsException) {
            // Can sometimes happen when absoluteX and absoluteY are called
        }
        return true
    }

    private fun twoStep(view: ZoomLayout, x: Float, y: Float) {
        if (view.scale > view.minScale) {
            view.setScale(view.minScale, true)
        } else {
            view.setScale(view.maxScale, x, y, true)
        }
    }

    private fun threeStep(view: ZoomLayout, x: Float, y: Float) {
        val scale = view.scale
        val medium = view.minScale + (view.maxScale - view.minScale) * 0.3f
        if (scale < medium) {
            view.setScale(medium, x, y, true)
        } else if (scale >= medium && scale < view.maxScale) {
            view.setScale(view.maxScale, x, y, true)
        } else {
            view.setScale(view.minScale, true)
        }
    }
}